File: src/modlunky2/category/chain/eggplant.py
import logging
from typing import Set

from modlunky2.category.chain.common import (
    ChainMixin,
    ChainStepper,
)
from modlunky2.mem.entities import EntityType
from modlunky2.mem.state import Screen, State, Theme

logger = logging.getLogger("modlunky2")


class EggplantChain(ChainMixin):
    @classmethod
    def make_stepper(cls) -> ChainStepper:
        return ChainStepper(cls.__name__, cls().collect_eggplant)

    def collect_eggplant(self, game_state: State, player_item_types: Set[EntityType]):
        if EntityType.ITEM_EGGPLANT in player_item_types:
            return self.in_progress(self.carry_eggplant_to_51)

        # It's possible to collect the child without the eggplant
        if game_state.world == 5 and self.some_companion_is(
            game_state, EntityType.CHAR_EGGPLANT_CHILD
        ):
            return self.in_progress(self.guide_eggplant_child_to_71)

        # It's possible a companion collected the eggplant
        if (
            game_state.screen is Screen.LEVEL_TRANSITION
            and self.some_companion_has_item(game_state, EntityType.ITEM_EGGPLANT)
        ):
            return self.in_progress(self.carry_eggplant_to_51)

        if game_state.world > 5:
            return self.failed()

        return self.unstarted()

    def carry_eggplant_to_51(
        self, game_state: State, player_item_types: Set[EntityType]
    ):
        if game_state.world > 4:
            return self.in_progress(self.collect_eggplant_child)

        if game_state.screen is not Screen.LEVEL_TRANSITION:
            return self.in_progress(self.carry_eggplant_to_51)

        if (
            EntityType.ITEM_EGGPLANT in player_item_types
            or EntityType.ITEM_EGGPLANT in game_state.waddler_storage
            or self.some_companion_has_item(game_state, EntityType.ITEM_EGGPLANT)
        ):
            return self.in_progress(self.carry_eggplant_to_51)

        # We can get another eggplant
        return self.unstarted()

    def collect_eggplant_child(self, game_state: State, _: Set[EntityType]):
        if self.some_companion_is(game_state, EntityType.CHAR_EGGPLANT_CHILD):
            return self.in_progress(self.guide_eggplant_child_to_71)

        if game_state.world > 5:
            return self.failed()

        return self.in_progress(self.collect_eggplant_child)

    def guide_eggplant_child_to_71(self, game_state: State, _: Set[EntityType]):
        world_level = (game_state.world, game_state.level)
        if world_level > (7, 1):
            return self.in_progress(self.visit_eggplant_world)

        if game_state.screen is not Screen.LEVEL_TRANSITION:
            return self.in_progress(self.guide_eggplant_child_to_71)

        # We won't have the child during the 7-1 transition
        if world_level == (7, 1):
            return self.in_progress(self.visit_eggplant_world)

        if self.some_companion_is(game_state, EntityType.CHAR_EGGPLANT_CHILD):
            return self.in_progress(self.guide_eggplant_child_to_71)

        return self.failed()

    def visit_eggplant_world(self, game_state: State, _: Set[EntityType]):
        if game_state.theme is Theme.EGGPLANT_WORLD:
            return self.in_progress(self.collect_eggplant_crown)

        if (game_state.world, game_state.level) > (7, 1):
            return self.failed()

        return self.in_progress(self.visit_eggplant_world)

    def collect_eggplant_crown(
        self, game_state: State, player_item_types: Set[EntityType]
    ):
        if EntityType.ITEM_POWERUP_EGGPLANTCROWN in player_item_types:
            return self.in_progress(self.success)

        if (game_state.world, game_state.level) > (7, 2):
            return self.failed()

        return self.in_progress(self.collect_eggplant_crown)

    def success(self, _unused1: State, _unused2: Set[EntityType]):
        # Eggplant can't fail once you have the crown
        return self.in_progress(self.success)

File: crates/ml2_trackers/src/chain_impl/eggplant.rs
//! Eggplant chain.
//!
//! Longer FSM: collect the eggplant, carry it to 5-1, hand it to the
//! Eggplant Child, guide the child through 7-1, visit Eggplant World,
//! collect the Eggplant Crown. Once the crown is in-hand the run
//! can't fall out of chain again.

use crate::chain::{ChainStepResult, ChainStepper};
use crate::chain_impl::inputs::ChainInputs;
use crate::entity_types as et;
use crate::enums::{Screen, Theme};

pub fn make_stepper() -> ChainStepper<ChainInputs> {
    ChainStepper::new("EggplantChain", collect_eggplant)
}

pub(crate) fn collect_eggplant(ctx: &ChainInputs) -> ChainStepResult<ChainInputs> {
    if ctx.player_items.contains(&et::ITEM_EGGPLANT) {
        return ChainStepResult::in_progress(carry_eggplant_to_51);
    }
    // The eggplant child can be picked up in world 5 even without
    // hand-carrying the eggplant (co-op edge case).
    if ctx.state.world == 5 && ctx.some_companion_is(et::CHAR_EGGPLANT_CHILD) {
        return ChainStepResult::in_progress(guide_eggplant_child_to_71);
    }
    // A companion may have grabbed the eggplant during the transition.
    if ctx.state.screen == Screen::LevelTransition && ctx.some_companion_has_item(et::ITEM_EGGPLANT)
    {
        return ChainStepResult::in_progress(carry_eggplant_to_51);
    }
    if ctx.state.world > 5 {
        return ChainStepResult::failed();
    }
    ChainStepResult::unstarted()
}

pub(crate) fn carry_eggplant_to_51(ctx: &ChainInputs) -> ChainStepResult<ChainInputs> {
    if ctx.state.world > 4 {
        return ChainStepResult::in_progress(collect_eggplant_child);
    }
    if ctx.state.screen != Screen::LevelTransition {
        return ChainStepResult::in_progress(carry_eggplant_to_51);
    }
    if ctx.player_items.contains(&et::ITEM_EGGPLANT)
        || ctx.waddler_contains(et::ITEM_EGGPLANT)
        || ctx.some_companion_has_item(et::ITEM_EGGPLANT)
    {
        return ChainStepResult::in_progress(carry_eggplant_to_51);
    }
    // Dropped the eggplant. Player can grab another this run, reset.
    ChainStepResult::unstarted()
}

pub(crate) fn collect_eggplant_child(ctx: &ChainInputs) -> ChainStepResult<ChainInputs> {
    if ctx.some_companion_is(et::CHAR_EGGPLANT_CHILD) {
        return ChainStepResult::in_progress(guide_eggplant_child_to_71);
    }
    if ctx.state.world > 5 {
        return ChainStepResult::failed();
    }
    ChainStepResult::in_progress(collect_eggplant_child)
}

pub(crate) fn guide_eggplant_child_to_71(ctx: &ChainInputs) -> ChainStepResult<ChainInputs> {
    let wl = ctx.state.world_level();
    if wl > (7, 1) {
        return ChainStepResult::in_progress(visit_eggplant_world);
    }
    if ctx.state.screen != Screen::LevelTransition {
        return ChainStepResult::in_progress(guide_eggplant_child_to_71);
    }
    // The child is transferred off-screen during the 7-1 transition,
    // so its absence there specifically doesn't fail the chain.
    if wl == (7, 1) {
        return ChainStepResult::in_progress(visit_eggplant_world);
    }
    if ctx.some_companion_is(et::CHAR_EGGPLANT_CHILD) {
        return ChainStepResult::in_progress(guide_eggplant_child_to_71);
    }
    ChainStepResult::failed()
}

pub(crate) fn visit_eggplant_world(ctx: &ChainInputs) -> ChainStepResult<ChainInputs> {
    if ctx.state.theme == Theme::EggplantWorld {
        return ChainStepResult::in_progress(collect_eggplant_crown);
    }
    if ctx.state.world_level() > (7, 1) {
        return ChainStepResult::failed();
    }
    ChainStepResult::in_progress(visit_eggplant_world)
}

pub(crate) fn collect_eggplant_crown(ctx: &ChainInputs) -> ChainStepResult<ChainInputs> {
    if ctx.player_items.contains(&et::ITEM_POWERUP_EGGPLANTCROWN) {
        return ChainStepResult::in_progress(success);
    }
    if ctx.state.world_level() > (7, 2) {
        return ChainStepResult::failed();
    }
    ChainStepResult::in_progress(collect_eggplant_crown)
}

pub(crate) fn success(_ctx: &ChainInputs) -> ChainStepResult<ChainInputs> {
    // Once the crown lands the chain can't fail. Perpetual
    // InProgress(success) is the fixed point.
    ChainStepResult::in_progress(success)
}
